File: android/feature/files/src/main/java/com/bytebox/feature/files/presentation/notifications/NotificationsViewModel.kt
package com.bytebox.feature.files.presentation.notifications

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bytebox.core.common.Result
import com.bytebox.domain.model.NotificationItem
import com.bytebox.domain.repository.FileRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class NotificationsUiState(
    val notifications: List<NotificationItem> = emptyList(),
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
    val unreadCount: Int = 0
)

@HiltViewModel
class NotificationsViewModel @Inject constructor(
    private val fileRepository: FileRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(NotificationsUiState())
    val uiState: StateFlow<NotificationsUiState> = _uiState.asStateFlow()

    init {
        loadNotifications()
    }

    fun loadNotifications() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            when (val result = fileRepository.getNotifications()) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        notifications = result.data,
                        isLoading = false,
                        unreadCount = result.data.count { !it.isRead }
                    )
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        errorMessage = result.exception.message
                    )
                }
                is Result.Loading -> {}
            }
        }
    }

    fun loadUnreadCount() {
        viewModelScope.launch {
            when (val result = fileRepository.getUnreadNotificationCount()) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(unreadCount = result.data)
                }
                else -> {}
            }
        }
    }

    fun markAsRead(id: String) {
        viewModelScope.launch {
            fileRepository.markNotificationRead(id)
            _uiState.value = _uiState.value.copy(
                notifications = _uiState.value.notifications.map {
                    if (it.id == id) it.copy(isRead = true) else it
                },
                unreadCount = (_uiState.value.unreadCount - 1).coerceAtLeast(0)
            )
        }
    }

    fun markAllAsRead() {
        viewModelScope.launch {
            fileRepository.markAllNotificationsRead()
            _uiState.value = _uiState.value.copy(
                notifications = _uiState.value.notifications.map { it.copy(isRead = true) },
                unreadCount = 0
            )
        }
    }
}

File: ios/ByteBox/ViewModels/NotificationsViewModel.swift
import Foundation

@MainActor
final class NotificationsViewModel: ObservableObject {
    @Published var notifications: [AppNotification] = []
    @Published var unreadCount: Int = 0
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let fileService = FileService.shared

    func loadNotifications() async {
        isLoading = true
        errorMessage = nil

        do {
            notifications = try await fileService.getNotifications()
            unreadCount = notifications.filter { !$0.isRead }.count
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    func loadUnreadCount() async {
        do {
            unreadCount = try await fileService.getUnreadCount()
        } catch {
            // Silently fail for badge count
        }
    }

    func markAsRead(_ id: String) async {
        do {
            try await fileService.markNotificationRead(id)
            if let index = notifications.firstIndex(where: { $0.id == id }) {
                let n = notifications[index]
                notifications[index] = AppNotification(
                    id: n.id,
                    type: n.type,
                    title: n.title,
                    message: n.message,
                    isRead: true,
                    createdAt: n.createdAt
                )
                unreadCount = notifications.filter { !$0.isRead }.count
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func markAllAsRead() async {
        do {
            try await fileService.markAllNotificationsRead()
            notifications = notifications.map {
                AppNotification(
                    id: $0.id,
                    type: $0.type,
                    title: $0.title,
                    message: $0.message,
                    isRead: true,
                    createdAt: $0.createdAt
                )
            }
            unreadCount = 0
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}
